crypto/pkcs7/bio/cipher.c

```diff
@@ -0,0 +1,320 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <openssl/buffer.h>
+#include <openssl/crypto.h>
+#include <openssl/evp.h>
+#include <openssl/mem.h>
+#include <openssl/pkcs7.h>
+#include <stdio.h>
+#include "../../fipsmodule/cipher/internal.h"
+#include "../internal.h"
+
+static int enc_write(BIO *h, const char *buf, int num);
+static int enc_read(BIO *h, char *buf, int size);
+static long enc_ctrl(BIO *h, int cmd, long arg1, void *arg2);
+static int enc_new(BIO *h);
+static int enc_free(BIO *data);
+#define ENC_BLOCK_SIZE (1024 * 4)
+
+typedef struct enc_struct {
+  uint8_t done;     // > 0 when finished
+  uint8_t flushed;  // > 0 when buffered data has been flushed
+  uint8_t ok;       // decryption status
+  int buf_off;      // start idx of buffered data
+  int buf_len;      // length of buffered data
+  EVP_CIPHER_CTX *cipher;
+  uint8_t buf[ENC_BLOCK_SIZE];
+} BIO_ENC_CTX;
+
+static const BIO_METHOD methods_enc = {
+    BIO_TYPE_CIPHER,
+    "cipher",
+    enc_write,
+    enc_read,
+    NULL,  // enc_puts
+    NULL,  // enc_gets
+    enc_ctrl,
+    enc_new,
+    enc_free,
+    NULL,  // enc_callback_ctrl
+};
+
+const BIO_METHOD *BIO_f_cipher(void) { return &methods_enc; }
+
+static int enc_new(BIO *b) {
+  BIO_ENC_CTX *ctx;
+  GUARD_PTR(b);
+
+  if ((ctx = OPENSSL_zalloc(sizeof(*ctx))) == NULL) {
+    return 0;
+  }
+
+  ctx->cipher = EVP_CIPHER_CTX_new();
+  if (ctx->cipher == NULL) {
+    OPENSSL_free(ctx);
+    return 0;
+  }
+  ctx->done = 0;
+  ctx->flushed = 0;
+  ctx->ok = 1;
+  ctx->buf_off = 0;
+  ctx->buf_len = 0;
+  BIO_set_data(b, ctx);
+  BIO_set_init(b, 1);
+
+  return 1;
+}
+
+static int enc_free(BIO *b) {
+  BIO_ENC_CTX *ctx;
+  GUARD_PTR(b);
+
+  ctx = BIO_get_data(b);
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  EVP_CIPHER_CTX_free(ctx->cipher);
+  OPENSSL_free(ctx);
+  BIO_set_data(b, NULL);
+  BIO_set_init(b, 0);
+
+  return 1;
+}
+
+static int enc_read(BIO *b, char *out, int outl) {
+  GUARD_PTR(b);
+  GUARD_PTR(out);
+  BIO_ENC_CTX *ctx = BIO_get_data(b);
+  if (ctx == NULL) {
+    return 0;
+  }
+  BIO *next = BIO_next(b);
+  if (next == NULL) {
+    return 0;
+  }
+
+  int bytes_processed = 0;
+  int remaining = outl;
+  const int cipher_block_size = EVP_CIPHER_CTX_block_size(ctx->cipher);
+  while (remaining > 0) {
+    if (ctx->buf_len > 0) {
+      int bytes_decrypted;
+      int to_decrypt = remaining > ctx->buf_len ? ctx->buf_len : remaining;
+      uint8_t *out_pos = ((uint8_t *)out) + bytes_processed;
+      // |EVP_DecryptUpdate| may write up to cipher_block_size-1 more bytes than
+      // requested, so return early if we cannot accommodate that with current
+      // |remaining| byte count.
+      if ((to_decrypt > (remaining - cipher_block_size + 1)) ||
```

Comment:
There's no guarantee that the caller will ever pass in an `outl` that's larger than `cipher_block_size + 1`.  For reads, I think `ctx-buf` should be a buffer of plaintext instead of the ciphertext.

Reply:
I've rewritten the read routine to buffer in plaintext, decrypting immediately after reading from the underlying BIO.